compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp

```diff
@@ -24,10 +27,274 @@
+  }
+};
+
+FailureOr<SmallVector<Operation *>>
+BridgeInParallelToFlowAbstractionGap::returningMatchAndRewrite(
+    LinalgExt::InParallelOp inParallelOp, PatternRewriter &rewriter) const {
+  if (inParallelOp->getNumResults() == 0) return SmallVector<Operation *>();
+
+  // If any result is not a single use tensors being tied back into a HAL
+  // subspan op, bail. This is not considered a failure.
+  // TODO: allow all or nothing, otherwise failure.
+  if (llvm::any_of(inParallelOp->getResults(), [](Value result) {
+        if (!result.hasOneUse()) return true;
+        auto storeOp =
+            dyn_cast<Flow::DispatchTensorStoreOp>(*(result.getUsers().begin()));
+        return !storeOp ||
+               !storeOp.target()
+                    .getDefiningOp<HAL::InterfaceBindingSubspanOp>();
+      }))
+    return SmallVector<Operation *>();
+
+  for (OpResult result : inParallelOp->getResults()) {
+    auto tensorStoreOp =
+        cast<Flow::DispatchTensorStoreOp>(*(result.getUsers().begin()));
+    auto subSpanOp =
+        tensorStoreOp.target().getDefiningOp<HAL::InterfaceBindingSubspanOp>();
+    LinalgExt::ParallelInsertSliceOp yieldingOp =
+        inParallelOp.getTerminator().yieldingOps()[result.getResultNumber()];
+    // Replace OpOperand by a clone of the HAL binding subspan op and forward
+    // the uses of flow.dispatch.tensor/load.
+    // TODO: check the load is of the full tensor, OTOH this whole process is a
+    // hack that needs to go away ...
+    SmallVector<Flow::DispatchTensorLoadOp> loadsFromSubspan;
+    for (Operation *op : subSpanOp.result().getUsers()) {
+      auto user = dyn_cast<Flow::DispatchTensorLoadOp>(op);
+      if (user) loadsFromSubspan.push_back(user);
+    }
+    rewriter.setInsertionPoint(subSpanOp);
+    // clang-format off
+    auto newSubspanOp =
+        rewriter.create<HAL::InterfaceBindingSubspanOp>(
+            inParallelOp->getLoc(),
+            yieldingOp.yieldedType(),
+            subSpanOp.set(),
+            subSpanOp.binding(),
+            subSpanOp.type(),
+            subSpanOp.byte_offset(),
+            subSpanOp.dynamic_dims(),
+            subSpanOp.alignment() ?
+              rewriter.getIndexAttr(subSpanOp.alignment()->getZExtValue()) :
+              IntegerAttr());
+    rewriter.replaceOpWithIf(subSpanOp, newSubspanOp.result(), [](OpOperand &operand){
+      return !isa<Flow::DispatchTensorLoadOp>(operand.getOwner());
+    });
+    // for (auto loadOp : loadsFromSubspan)
+    //   rewriter.replaceOp(loadOp, newSubspanOp.result());
+    // clang-format on
+
+    yieldingOp.destMutable().assign(newSubspanOp.getResult());
+    rewriter.eraseOp(tensorStoreOp);
+  }
+
+  // All results are assumed tied back into HAL. This serves purely as a DCE
+  // avoidance mechanism.
+  rewriter.setInsertionPoint(inParallelOp->getBlock()->getTerminator());
+  rewriter.create<LinalgExt::AssumeTiedToHALOp>(inParallelOp->getLoc(),
```

Comment:
This seems like a strange construct just to avoid using the `flow.dispatch.tensor.store`. Using the SSA value to signal something is written out is the current way things are setup. It would be good to stay consistent with that.